pkg/forklift-api/webhooks/mutating-webhook/mutators/plan-mutator.go

```diff
@@ -90,55 +127,72 @@ func (mutator *PlanMutator) Mutate(ar *admissionv1.AdmissionReview) *admissionv1
-					return util.ToAdmissionResponseError(err)
+					return false, err
 				}
 				tcl, err = targetProvider.Client(secret)
 			}
 			if err != nil {
 				log.Error(err, "Failed to initiate client to target cluster", err.Error())
-				return util.ToAdmissionResponseError(err)
+				return false, err
 			}
 
 			netAttachDef := &net.NetworkAttachmentDefinition{}
 			if err = tcl.Get(context.TODO(), key, netAttachDef); err == nil {
 				log.Info("Patching the plan's transfer network")
-				plan.Spec.TransferNetwork = &core.ObjectReference{
+				mutator.plan.Spec.TransferNetwork = &core.ObjectReference{
 					Name:      network,
-					Namespace: plan.Spec.TargetNamespace,
+					Namespace: mutator.plan.Spec.TargetNamespace,
 				}
 				planChanged = true
 			} else if !k8serr.IsNotFound(err) { // TODO: else if !NotFound ...
 				log.Error(err, "Failed to get the network-attachment-definition", err.Error())
-				return util.ToAdmissionResponseError(err)
+				return false, err
 			}
 		}
 	}
+	return planChanged, nil
+}
 
-	if planChanged {
-		patchBytes, err := util.GeneratePatchPayload(util.PatchOperation{
-			Op:    "replace",
-			Path:  "/spec",
-			Value: plan.Spec,
-		})
-
-		if err != nil {
-			log.Error(err, "mutating webhook error, failed to generete paylod for patch request")
-			return util.ToAdmissionResponseError(err)
+func (mutator *PlanMutator) populatorLabelsAnnotation() bool {
+	if mutator.ar.Request.Operation == admissionv1.Create {
+		if _, ok := mutator.plan.Annotations[AnnPopulatorLabels]; !ok {
+			if mutator.plan.Annotations == nil {
+				mutator.plan.Annotations = make(map[string]string)
+			}
+			mutator.plan.Annotations[AnnPopulatorLabels] = "True"
+			log.Info("Patching the plan's annotation")
+			return true
 		}
+	}
+	return false
+}
 
-		jsonPatchType := admissionv1.PatchTypeJSONPatch
-		return &admissionv1.AdmissionResponse{
-			Allowed:   true,
-			Patch:     patchBytes,
-			PatchType: &jsonPatchType,
-		}
-	} else {
-		return &admissionv1.AdmissionResponse{
-			Allowed: true,
-			Result: &metav1.Status{
-				Message: "Certificate retrieval is not required, passing ",
-				Code:    http.StatusOK,
+func (mutator *PlanMutator) patchPayload(specChanged, metadataChanged bool) ([]byte, error) {
+	if specChanged && metadataChanged {
+		return util.GeneratePatchPayload(
+			util.PatchOperation{
+				Op:    "replace",
+				Path:  "/spec",
+				Value: mutator.plan.Spec,
 			},
-		}
+			util.PatchOperation{
+				Op:    "replace",
+				Path:  "/metadata",
+				Value: mutator.plan.ObjectMeta,
+			},
+		)
+	} else if specChanged {
+		return util.GeneratePatchPayload(util.PatchOperation{
+			Op:    "replace",
+			Path:  "/spec",
+			Value: mutator.plan.Spec,
+		})
+	} else if metadataChanged {
+		return util.GeneratePatchPayload(util.PatchOperation{
+			Op:    "replace",
+			Path:  "/metadata",
+			Value: mutator.plan.ObjectMeta,
+		})
 	}
+	return nil, fmt.Errorf("no change, although should be one")
```

Comment:
this function is over-complicated - there are three cases above instead of two and it returns an error that cannot happen. how about returning an array of PatchOperation instead and then the caller can check if it's empty or not - if it's empty then return 'allow' without patching and otherwise patch the plan? I mean something similar to the code we have in secret-mutator

Reply:
the caller already checks is spec/metadata changed if not it will allow. as the previous logic was. but regarding making it easier with array - yes make sense.